### PR TITLE
Parse Srcset Attribute

### PR DESCRIPTION
**Pull request by iamneha**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #__ (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18313)

### CODE REVIEW

File: components/script/dom/htmlimageelement.rs [hunk 1]
```diff
@@ -1032,3 +1053,228 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                            if !current_descriptor.is_empty() {
+                                descriptors.push_back(current_descriptor.clone());
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        let mut error = Error::No;
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: This could simply be a `bool`, since there are only two possible values, and we can use bools easily in `if` conditionals.

File: components/script/dom/htmlimageelement.rs [hunk 2]
```diff
@@ -1032,3 +1053,228 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        let mut width: Option<u32> = None;
+        let mut density: Option<f64> = None;
+        let mut future_compat_h: Option<u32> = None;
+        for descriptor in descriptors {
+            let mut char_iter = descriptor.chars();
+            let (digits, remaining) = collect_sequence_characters(&descriptor, is_ascii_digit);
+            let valid_non_negative_integer = parse_unsigned_integer(digits.chars());
+            let has_w = if let "w" = remaining {
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: Instead of having a complicated pattern match and a variable binding, we should use `remaining == "w"` in the `if` conditionals below.

File: components/script/dom/htmlimageelement.rs [hunk 3]
```diff
@@ -1032,3 +1053,228 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                false
+            };
+            let has_h = if let "h" = remaining {
+                true
+            } else {
+                false
+            };
+            if valid_non_negative_integer.is_ok() && has_w {
+                //not support sizes attribute
+                if width != None && density != None {
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: Nit: Let's use `is_some()` and `is_none()` here and everywhere (appropriately). Note that there are plenty of useful methods for `Option<T>` and `Result<T, E>`

File: components/script/dom/htmlimageelement.rs [hunk 4]
```diff
@@ -1032,3 +1053,228 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                    error = Error::Yes;
+                }
+                let result = parse_unsigned_integer(char_iter.clone());
+                if result.is_err() {
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: If error becomes a bool, then we could write,

``` rust
error = result.is_err();
if let Ok(w) = result {
    width = Some(w);
}
```

File: components/script/dom/htmlimageelement.rs [hunk 5]
```diff
@@ -1032,3 +1053,228 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+
+        let(spaces, position) = collect_sequence_characters(position, |c| *c ==',' || char::is_whitespace(*c));
+        let x = spaces.find(',');
+        match x {
+            Some(val) => println!("Parse Error"),
+            None => println!("No commas\n"),
+        }
+
+        // add the counts of spaces that we collect to advance the start index
+        let space_len = spaces.char_indices().count();
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: We should go for `chars` only if we're concerned with unicode codepoints. If it's just spaces, then we could simply use `spaces.len()`

File: components/script/dom/htmlimageelement.rs [hunk 6]
```diff
@@ -1032,3 +1053,228 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        start += space_len;
+
+        //Returns and breaks out of the loop
+        if position.is_empty() {
+            return candidates;
+        }
+        let(url, spaces) = collect_sequence_characters(position, |c| !char::is_whitespace(*c));
+
+        // add the counts of urls that we parse to advance the start index
+        start += url.char_indices().count();
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: Why should we use `char_indices` instead of `chars` here? (and below?). We should prefer the latter, unless we're using the indices during the iteration.

File: components/script/dom/htmlimageelement.rs [hunk 7]
```diff
@@ -1032,3 +1053,228 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                        Some((idx, c @ ' ')) => {
+                            if !current_descriptor.is_empty() {
+                                descriptors.push_back(current_descriptor.clone());
+                                current_descriptor = String::new();
+                                state = ParseState::AfterDescriptor;
+                            }
+                            continue;
+                        }
+                        Some((idx, c @ ',')) => {
+                            position.chars().enumerate();
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: What's this?

**Review by ghost — CHANGES_REQUESTED** [review]
Just had a quick look. This looks good so far. Nice work! But, this still requires a lot of tidying up. Please fix the build warnings and run `./mach test-tidy` to help cleanup the code.

**Comment by wafflespeanut** [comment]
Also, please leave steps from spec in comments, so that it's easier to address/change parts of code later.

File: components/script/dom/htmlimageelement.rs [hunk 8]
```diff
@@ -1032,3 +1053,228 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                true
+            } else {
+                false
+            };
+            let has_h = if let "h" = remaining {
+                true
+            } else {
+                false
+            };
+            if valid_non_negative_integer.is_ok() && has_w {
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: During descriptor parsing, we ensure that there are no repetitions in descriptors, which means there shouldn't be any `w`, `x` or `h` while parsing the values - i.e., if the value ends with `h` and we already have a height value in the descriptor, then it's a parse error.

File: components/script/dom/htmlimageelement.rs [hunk 9]
```diff
@@ -1032,3 +1053,228 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+pub fn parse_a_srcset_attribute(input: String) -> Vec<ImageSource> {
+    let mut start = 0;
+    let mut candidates: Vec<ImageSource> = Vec::new();
+    loop {
+        let position = &input[start..];
+
+        let(spaces, position) = collect_sequence_characters(position, |c| *c ==',' || char::is_whitespace(*c));
+        let x = spaces.find(',');
+        match x {
+            Some(val) => println!("Parse Error"),
```
[the remark this answers is not in the document]
Comment (iamneha) [reply components/script/dom/htmlimageelement.rs]: Currently being used solely for debugging purpose and will be removed when PR is finalized

**Comment by wafflespeanut** [comment]
Step 13 [in the spec](https://html.spec.whatwg.org/multipage/images.html#parsing-a-srcset-attribute) looks funny.

> If the descriptor consists of a valid non-negative integer followed by a U+0068 LATIN SMALL LETTER H character

>    This is a parse error.

If it's a parse error, then there's no reason to parse it. The value for *future-compat-h* is not being used after parsing (probably because it's planned for the future?), but that shouldn't mean that the entire descriptor should be rejected with a parse error. We should probably skip the value and still use the descriptor?

/cc @jdm

**Comment by KiChjang** [comment]
Sorry, but which issue does this PR solve?

**Comment by iamneha** [comment]
@KiChjang we are solving srcset attribute from issue Support responsive images #11416

File: components/style/str.rs [hunk 10]
```diff
@@ -134,6 +135,29 @@ pub fn read_exponent<I: Iterator<Item=char>>(mut iter: Peekable<I>) -> Option<i3
     }
 }
 
+struct PeekableTakeWhile<'a, I: Iterator + 'a, F>(&'a mut ::std::iter::Peekable<I>, F, bool);
```
Comment (creativcoder) [inline components/style/str.rs]: Are we using these anywhere now ? If not then we can remove this and below `impl` block for it

File: components/script/dom/htmlimageelement.rs [hunk 11]
```diff
@@ -1032,3 +1052,216 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+            let char_iter = descriptor.chars();
+            let (digits, remaining) = collect_sequence_characters(&descriptor, is_ascii_digit);
+            let valid_non_negative_integer = parse_unsigned_integer(digits.chars());
+            let has_w = if remaining == "w" {
+                true
+            } else {
+                false
+            };
+            let valid_floating_point = parse_double(digits);
+            let has_x = if remaining == "x" {
```
Comment (creativcoder) [inline components/script/dom/htmlimageelement.rs]: we can just do this: `let has_x = remaining == "x"`

File: components/script/dom/htmlimageelement.rs [hunk 12]
```diff
@@ -1032,3 +1052,216 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                true
+            } else {
+                false
+            };
+            let has_h = if remaining == "h" {
```
Comment (creativcoder) [inline components/script/dom/htmlimageelement.rs]: Same as above

File: components/script/dom/htmlimageelement.rs [hunk 13]
```diff
@@ -1032,3 +1052,216 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                true
+            } else {
+                false
+            };
+            if valid_non_negative_integer.is_ok() && has_w {
+                //not support sizes attribute
+                if width.is_some() && density.is_some() {
```
Comment (creativcoder) [inline components/script/dom/htmlimageelement.rs]: you can do this `error = width.is_some() && density.is_some();`

File: components/script/dom/htmlimageelement.rs [hunk 14]
```diff
@@ -1032,3 +1052,216 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                    error = true;
+                }
+                let result = parse_unsigned_integer(char_iter.clone());
+                error = result.is_err();
+                if let Ok(w) = result {
```
Comment (creativcoder) [inline components/script/dom/htmlimageelement.rs]: As you are just pulling the value out and converting it to option type (i.e, `Some(w)`), you can use the ok method like: `width = result.ok()`. Calling `ok()` on a `Result<T>` type converts it to an `Option<T>`. `ok()` method does what you are currently doing here.

File: components/script/dom/htmlimageelement.rs [hunk 15]
```diff
@@ -1032,3 +1052,216 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                    width = Some(w);
```
Comment (creativcoder) [inline components/script/dom/htmlimageelement.rs]: As you are just pulling the value out and converting it to option type (i.e, `Some(w)`), you can use the ok method like: `width = result.ok()`. Calling `ok()` on a `Result<T>` type converts it to an `Option<T>`. `ok()` method does what you are currently doing here.

File: components/script/dom/htmlimageelement.rs [hunk 16]
```diff
@@ -1032,3 +1052,216 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                    future_compat_h = Some(h);
+                }
+            } else {
+                error = true;
+            }
+        }
+        if future_compat_h.is_some() && width.is_none() {
+            error = true;
+        }
+        if error == false {
```
Comment (creativcoder) [inline components/script/dom/htmlimageelement.rs]: As error now is just a boolean we can reduce the conditional check to just this:
```rust
if !error {
// stuff
}
```

File: components/script/dom/htmlimageelement.rs [hunk 17]
```diff
@@ -1032,3 +1052,216 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        }
+        let mut error = false;
+        let mut width: Option<u32> = None;
+        let mut density: Option<f64> = None;
+        let mut future_compat_h: Option<u32> = None;
+        for descriptor in descriptors {
+            let char_iter = descriptor.chars();
+            let (digits, remaining) = collect_sequence_characters(&descriptor, is_ascii_digit);
+            let valid_non_negative_integer = parse_unsigned_integer(digits.chars());
+            let has_w = if remaining == "w" {
```
Comment (creativcoder) [inline components/script/dom/htmlimageelement.rs]: we can just do this: `let has_w = remaining == "w"`

File: components/script/dom/webidls/HTMLImageElement.webidl [hunk 18]
```diff
@@ -10,7 +10,7 @@ interface HTMLImageElement : HTMLElement {
   [CEReactions]
            attribute DOMString src;
   // [CEReactions]
-  //         attribute DOMString srcset;
```
Comment (creativcoder) [inline components/script/dom/webidls/HTMLImageElement.webidl]: Nit: Please remove this commented line

File: components/script/dom/htmlimageelement.rs [hunk 19]
```diff
@@ -1032,3 +1052,216 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        // add 1 to start index, for the comma
+        start += 1;
+        let(space, position) = collect_sequence_characters(spaces, |c| char::is_whitespace(*c));
+        let space_len = space.len();
+        start += space_len;
+        let mut descriptors = LinkedList::<String>::new();
+        let mut current_descriptor = String::new();
+        let mut state = ParseState::InDescriptor;
+        let mut char_stream = position.chars().enumerate();
+        //let mut last_index = 0;
```
Comment (creativcoder) [inline components/script/dom/htmlimageelement.rs]: Nit: Please remove the commented line

File: components/script/dom/htmlimageelement.rs [hunk 20]
```diff
@@ -1032,3 +1052,216 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                }
+            } else {
+                error = true;
+            }
+        }
+        if future_compat_h.is_some() && width.is_none() {
+            error = true;
+        }
+        if error == false {
+            if width.is_some() || density.is_some() {
```
Comment (creativcoder) [inline components/script/dom/htmlimageelement.rs]: According to Step 15 of [spec](https://html.spec.whatwg.org/multipage/images.html#parsing-a-srcset-attribute): is written in a misleading way. Due to this check, the case of a single srcset url without any descriptor/density : (`small_image.jpg`), does not parses correctly and returns an empty vec. This is a valid srcset string as confirmed by neha.

The spec does not mean that we should append Image Source only if width and density is present.
Rather it should just check if error is false and just append whatever is there in the `width` and `density` variable along with url.

So basically this : 
```rust
if !error {
// append stuff to candidates
} else {
println!("parsed error");
}
```

File: components/script/dom/htmlimageelement.rs [hunk 21]
```diff
@@ -1032,3 +1052,199 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+
+pub fn parse_a_srcset_attribute(input: String) -> Vec<ImageSource> {
+    let mut start = 0;
+    let mut candidates: Vec<ImageSource> = Vec::new();
+    while start < input.len() {
+        let position = &input[start..];
+        let(spaces, position) = collect_sequence_characters(position, |c| *c ==',' || char::is_whitespace(*c));
+        let x = spaces.find(',');
+        match x {
+            Some(val) => println!("Parse Error"),
```
Comment (KiChjang) [inline components/script/dom/htmlimageelement.rs]: As previously stated, you should return an Error here instead of `println!`.

File: components/script/dom/webidls/HTMLImageElement.webidl [hunk 22]
```diff
@@ -9,8 +9,8 @@ interface HTMLImageElement : HTMLElement {
            attribute DOMString alt;
   [CEReactions]
            attribute DOMString src;
-  // [CEReactions]
-  //         attribute DOMString srcset;
+  [CEReactions]
```
Comment (KiChjang) [inline components/script/dom/webidls/HTMLImageElement.webidl]: In order to generate appropriate bindings for the return type, you'll need to add the `[Throws]` attribute here.

File: components/script/test.rs [hunk 23]
```diff
@@ -10,6 +10,7 @@ pub use dom::bindings::cell::DOMRefCell;
 pub use dom::bindings::js::JS;
 pub use dom::node::Node;
 pub use dom::bindings::refcounted::TrustedPromise;
+pub use std::string::String;
```
Comment (KiChjang) [inline components/script/test.rs]: This should not be necessary.

File: tests/unit/script/htmlimageelement.rs [hunk 24]
```diff
@@ -0,0 +1,72 @@
+use script::test::String;
```
Comment (KiChjang) [inline tests/unit/script/htmlimageelement.rs]: This should not be necessary.

**Review by KiChjang — CHANGES_REQUESTED** [review]
Made a preliminary review, I have not gotten too far into the spec before I started finding problems.

It would also be super helpful if you can add step annotations inside your function body, this will leave a good note of which part of the spec that the code is attempting to follow.

File: components/script/dom/htmlimageelement.rs [hunk 25]
```diff
@@ -1032,3 +1052,199 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        {
+            for (i, ch) in s.chars().enumerate() {
+                if !predicate(&ch) {
+                    return (&s[0..i], &s[i..])
+                }
+            }
+        return (s, "");
+        }
+
+pub fn parse_a_srcset_attribute(input: String) -> Vec<ImageSource> {
```
[the remark this answers is not in the document]
Comment (KiChjang) [reply components/script/dom/htmlimageelement.rs]: Should also link to https://html.spec.whatwg.org/multipage/#parsing-a-srcset-attribute.

**Comment by jdm** [comment]
Stealing this back. Thanks for the enthusiasm, but there are already several people looking at this PR.

File: components/script/dom/htmlimageelement.rs (hunk 21 above)
Comment (jdm) [reply components/script/dom/htmlimageelement.rs]: As previously discussed with Neha, the spec does not actually want us to abort parsing when a parse error is encountered.

**Comment by jdm** [comment]
As previously discussed on IRC, I would expect to see tests for densities that have decimal values.

File: components/script/dom/htmlimageelement.rs [hunk 26]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        let url: String = url.chars().take(url.chars().count() - comma_count).collect();
+        if comma_count > 1 {
+            println!("Parse Error (trailing commas)")
+        }
+        // add 1 to start index, for the comma
+        start += 1;
+        let(space, position) = collect_sequence_characters(spaces, |c| char::is_whitespace(*c));
+        let space_len = space.len();
+        start += space_len;
+        let mut descriptors = LinkedList::<String>::new();
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: We can use a vector instead of a linked list here.

File: components/script/dom/htmlimageelement.rs [hunk 27]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        }
+        // add the counts of spaces that we collect to advance the start index
+        let space_len = spaces.char_indices().count();
+        start += space_len;
+        //Returns and breaks out of the loop
+        if position.is_empty() {
+            return candidates;
+        }
+        let(url, spaces) = collect_sequence_characters(position, |c| !char::is_whitespace(*c));
+        // add the counts of urls that we parse to advance the start index
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: `add the length of the url that we parse to advance the start index`

File: components/script/dom/htmlimageelement.rs [hunk 28]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+pub fn parse_a_srcset_attribute(input: String) -> Vec<ImageSource> {
+    let mut start = 0;
+    let mut candidates: Vec<ImageSource> = vec![];
+    while start < input.len() {
+        let position = &input[start..];
+        let(spaces, position) = collect_sequence_characters(position, |c| *c ==',' || char::is_whitespace(*c));
+        let x = spaces.find(',');
+        match x {
+            Some(val) => println!("Parse Error"),
+            None => println!("No commas\n"),
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: We should remove the println! calls now. Since a parse error does not actually change the way this code behaves, we can remove the check for the commas here.

File: components/script/dom/htmlimageelement.rs [hunk 29]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        if position.is_empty() {
+            return candidates;
+        }
+        let(url, spaces) = collect_sequence_characters(position, |c| !char::is_whitespace(*c));
+        // add the counts of urls that we parse to advance the start index
+        start += url.chars().count();
+        let comma_count = url.chars().rev().take_while(|c| *c == ',').count();
+        let url: String = url.chars().take(url.chars().count() - comma_count).collect();
+        if comma_count > 1 {
+            println!("Parse Error (trailing commas)")
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: Let's remove this println and the check, since a parse error does not affect our behaviour.

File: components/script/dom/htmlimageelement.rs [hunk 30]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        }
+        // add 1 to start index, for the comma
+        start += 1;
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: This should be increased by comma_count.

File: components/script/dom/htmlimageelement.rs [hunk 31]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        let(space, position) = collect_sequence_characters(spaces, |c| char::is_whitespace(*c));
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: nit: space after let.

File: components/script/dom/htmlimageelement.rs (hunk 7 above)
Comment (jdm) [reply components/script/dom/htmlimageelement.rs]: This should be removed, since it does not do anything.

File: components/script/dom/htmlimageelement.rs [hunk 32]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        let mut buffered: Option<(usize, char)> = None;
+        loop {
+            let nextChar = buffered.take().or_else(|| char_stream.next());
+            if let Some((i, _)) = nextChar {
+                start += 1;
+            }
+            match state {
+                ParseState::InDescriptor => {
+                    match nextChar {
+                        Some((idx, c @ ' ')) => {
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: No need for `c @` here since we don't use `c`.

File: components/script/dom/htmlimageelement.rs [hunk 33]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+            }
+        }
+        let mut error = false;
+        let mut width: Option<u32> = None;
+        let mut density: Option<f64> = None;
+        let mut future_compat_h: Option<u32> = None;
+        for descriptor in descriptors {
+            let char_iter = descriptor.chars();
+            let (digits, remaining) = collect_sequence_characters(&descriptor, |c| is_ascii_digit(c) || *c == '.');
+            let valid_non_negative_integer = parse_unsigned_integer(digits.chars());
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: We probably don't need `.chars()` here.

File: components/script/dom/htmlimageelement.rs [hunk 34]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        let mut error = false;
+        let mut width: Option<u32> = None;
+        let mut density: Option<f64> = None;
+        let mut future_compat_h: Option<u32> = None;
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: We should declare these inside the loop, or else they will never be reset. I suspect that parsing `foo.jpg 2w, bar.jpg 1.0x` would fail right now.

File: components/script/dom/htmlimageelement.rs [hunk 35]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+            let has_x = remaining == "x";
+            let has_h = remaining == "h";
+            if valid_non_negative_integer.is_ok() && has_w {
+                let result = parse_unsigned_integer(char_iter.clone());
+                error = result.is_err();
+                if let Ok(w) = result {
+                    width = Some(w);
+                }
+            } else if valid_floating_point.is_ok() && has_x {
+                let result = parse_double(digits);
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: This should use char_iter as well.

File: components/script/dom/htmlimageelement.rs [hunk 36]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                let result = parse_unsigned_integer(char_iter.clone());
+                error = result.is_err();
+                if let Ok(h) = result {
+                    future_compat_h = Some(h);
+                }
+            } else {
+                error = true;
+            }
+            if (width.is_some() && density.is_some()) ||
+                (width.is_some() && density.is_some() && future_compat_h.is_some()) ||
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: These checks belong in the individual branches for each kind of descriptor. Right now, this code does not treat `foo.jpg 2.2x 1.1x` as an error, but it should be caught by "If width, density and future-compat-h are not all absent, then let error be yes."

File: components/script/dom/htmlimageelement.rs [hunk 37]
```diff
@@ -1032,3 +1047,197 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        }
+        if future_compat_h.is_some() && width.is_none() {
+            error = true;
+        }
+        if !error {
+            let descriptor = Descriptor { wid: width, den: density };
+            let imageSource = ImageSource { url: url, descriptor: descriptor };
+            candidates.push(imageSource);
+        } else {
+            debug!("Parse error")
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: Let's remove this else block now.

File: components/style/str.rs [hunk 38]
```diff
@@ -57,7 +57,8 @@ pub fn split_commas<'a>(s: &'a str) -> Filter<Split<'a, char>, fn(&&str) -> bool
     s.split(',').filter(not_empty as fn(&&str) -> bool)
 }
 
-fn is_ascii_digit(c: &char) -> bool {
+/// ascii digit
```
Comment (jdm) [inline components/style/str.rs]: `/// Is the character an ascii digit?`

File: components/script/test.rs (hunk 23 above)
Comment (jdm) [reply components/script/test.rs]: Let's remove this.

File: tests/unit/script/htmlimageelement.rs (hunk 24 above)
Comment (jdm) [reply tests/unit/script/htmlimageelement.rs]: Let's remove this.

File: tests/unit/script/htmlimageelement.rs [hunk 39]
```diff
@@ -0,0 +1,81 @@
...
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: This file is missing a license header.

File: tests/unit/script/htmlimageelement.rs [hunk 40]
```diff
@@ -0,0 +1,81 @@
...
+use script::test::srcset::{Descriptor, ImageSource, parse_a_srcset_attribute};
+
+#[test]
+fn no_value() {
+    let new_vec = Vec::new();
+    assert_eq!(parse_a_srcset_attribute(String::new()), new_vec);
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: We can use "" instead of String::new.

File: tests/unit/script/htmlimageelement.rs [hunk 41]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+#[test]
+fn width_one_value() {
+    let first_descriptor = Descriptor { wid: Some(320), den: None };
+    let first_imagesource = ImageSource { url: "small-image.jpg".to_string(), descriptor: first_descriptor };
+    let mut sources = Vec::new();
+    sources.push(first_imagesource);
+    assert_eq!(parse_a_srcset_attribute(String::from("small-image.jpg, 320w")), sources);
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: All of these tests can use string literals instead of String::from.

File: tests/unit/script/htmlimageelement.rs [hunk 42]
```diff
@@ -0,0 +1,81 @@
...
+    let new_vec = Vec::new();
+    assert_eq!(parse_a_srcset_attribute(String::new()), new_vec);
+}
+
+#[test]
+fn width_one_value() {
+    let first_descriptor = Descriptor { wid: Some(320), den: None };
+    let first_imagesource = ImageSource { url: "small-image.jpg".to_string(), descriptor: first_descriptor };
+    let mut sources = Vec::new();
+    sources.push(first_imagesource);
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: All of these tests can use `let sources = &[first_imagesource];` instead of pushing entries individually.

File: tests/unit/script/htmlimageelement.rs [hunk 43]
```diff
@@ -0,0 +1,81 @@
...
+    assert_eq!(parse_a_srcset_attribute(String::from("small-image.jpg")), sources);
+}
+//Does not parse an ImageSource when both width and density descriptor present
+#[test]
+fn two_descriptor() {
+    let first_descriptor = Descriptor { wid: Some(380), den: Some(22.0) };
+    let first_imagesource = ImageSource { url: "small-image.jpg".to_string(), descriptor: first_descriptor };
+    let mut sources = Vec::new();
+    sources.push(first_imagesource);
+    assert_ne!(parse_a_srcset_attribute(String::from("small-image.jpg 380w 22x")), sources);
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: Instead of checking that it does not equal this, let's check that it does equal the empty vector.

**Review by jdm — CHANGES_REQUESTED** [review]
Looking good!

File: components/script/dom/htmlimageelement.rs (hunk 34 above)
Comment (iamneha) [reply components/script/dom/htmlimageelement.rs]: It is not failing I tried this also and added test for this also

**Comment by KiChjang** [comment]
Again, please remove the merge commit.

File: components/script/dom/htmlimageelement.rs [hunk 44]
```diff
@@ -978,3 +1004,177 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        for descriptor in descriptors {
+            let char_iter = descriptor.chars();
+            let (digits, remaining) = collect_sequence_characters(&descriptor, |c| is_ascii_digit(c) || *c == '.');
+            let valid_non_negative_integer = parse_unsigned_integer(digits.chars());
+            let has_w = remaining == "w";
+            let valid_floating_point = parse_double(digits);
+            let has_x = remaining == "x";
+            let has_h = remaining == "h";
+            if valid_non_negative_integer.is_ok() && has_w {
+                let result = parse_unsigned_integer(char_iter.clone());
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: Why should we parse this again if we already have the result in `valid_non_negative_integer`? We could just `unwrap` it, right?

File: components/script/dom/htmlimageelement.rs (hunk 35 above)
Comment (wafflespeanut) [reply components/script/dom/htmlimageelement.rs]: Ditto on unnecessary re-parsing. We should unwrap from `valid_floating_point`

File: components/script/dom/htmlimageelement.rs [hunk 45]
```diff
@@ -978,3 +1004,177 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+                let result = parse_double(digits);
+                error = result.is_err();
+                if width.is_some() || density.is_some() || future_compat_h.is_some() {
+                    error = true;
+                }
+                if let Ok(x) = result {
+                    density = Some(x);
+                }
+            } else if valid_non_negative_integer.is_ok() && has_h {
+                let result = parse_unsigned_integer(char_iter.clone());
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: Ditto on re-parsing.

**Review by ghost — CHANGES_REQUESTED** [review]
Going good so far. Nice work! A few more steps towards merging. I think some of my comments should be thrown as warnings by rustc. We should get rid of the warnings just as we do the errors.

File: components/script/dom/htmlimageelement.rs [hunk 46]
```diff
@@ -978,3 +1002,176 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+pub fn collect_sequence_characters<F>(s: &str, predicate: F) -> (&str, &str)
+where F: Fn(&char) -> bool
+    {
+        for (i, ch) in s.chars().enumerate() {
+            if !predicate(&ch) {
+                return (&s[0..i], &s[i..])
+            }
+        }
+    return (s, "");
+    }
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: Here's the correct indentation:

``` rust
pub fn collect_sequence_characters<F>(s: &str, predicate: F) -> (&str, &str)
    where F: Fn(&char) -> bool
{
    for (i, ch) in s.chars().enumerate() {
        if !predicate(&ch) {
            return (&s[0..i], &s[i..])
        }
    }
    
    return (s, "");
}
```

File: components/script/dom/htmlimageelement.rs [hunk 47]
```diff
@@ -978,3 +1004,177 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
...
+        // add the length of the url that we parse to advance the start index
+        let space_len = spaces.char_indices().count();
+        start += space_len;
+        //Returns and breaks out of the loop
+        if position.is_empty() {
+            return candidates;
+        }
+        let (url, spaces) = collect_sequence_characters(position, |c| !char::is_whitespace(*c));
+        // add the counts of urls that we parse to advance the start index
+        start += url.chars().count();
```
[the remark this answers is not in the document]
Comment (wafflespeanut) [reply components/script/dom/htmlimageelement.rs]: This hasn't been addressed yet.

**Comment by wafflespeanut** [comment]
Sorry, I just realized that you haven't finished updating the code. 😅

**Comment by jdm** [comment]
I've squashed these commits in #18716.
